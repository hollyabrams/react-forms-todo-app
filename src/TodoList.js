import React, { useState, useEffect } from 'react';
import Todo from './Todo';
import NewTodoForm from './NewTodoForm';
import './TodoList.scss';

function TodoList() {
  const [todos, setTodos] = useState([]);

  // Get todos from localStorage on page load
  useEffect(() => {
    const storedTodos = JSON.parse(localStorage.getItem('todos'));
    if (storedTodos) {
      setTodos(storedTodos);
    }
  }, []);

  // Save todos to localStorage on state change
  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const addTodo = (newTodo) => {
    setTodos(todos => [...todos, newTodo]);
  };

  const removeTodo = (id) => {
    setTodos(todos => todos.filter(todo => todo.id !== id));
  };

  const removeAllTodos = () => {
    setTodos([]);
  };

  const editTodo = (id, newTask) => {
    setTodos(todos => todos.map(todo => {
      if (todo.id === id) {
        return { ...todo, task: newTask };
      }
      return todo;
    }));
  };

  const toggleCompletion = (id) => {
    setTodos(todos => todos.map(todo => {
      if (todo.id === id) {
        return { ...todo, completed: !todo.completed };
      }
      return todo;
    }));
  };

  return (
    <div className="container">
      <div className="todo-header">
        <h1 className="todo-header__title">Todo List</h1>
        <button
          onClick={removeAllTodos}
          className="todo-header__delete-all-button"
        >
          Delete All
        </button>
      </div>
      <NewTodoForm addTodo={addTodo} />
      <ul className="list-group">
        {todos.map(todo => (
          <Todo
            key={todo.id}
            id={todo.id}
            task={todo.task}
            completed={todo.completed}
            removeTodo={removeTodo}
            editTodo={editTodo}
            toggleCompletion={toggleCompletion}
          />
        ))}
      </ul>
    </div>
  );
}

export default TodoList;
