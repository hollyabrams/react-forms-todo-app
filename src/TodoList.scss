// variables
$primary-color: #3f51b5;
$danger-color: #f44336;
$success-color: #4caf50;
$dark-color: #212121;
$light-color: #f5f5f5;

// mixins
@mixin center-flex() {
    display: flex;
    justify-content: center;
    align-items: center;
}

// global styles
* {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
}

body {
    font-family: Arial, sans-serif;
    font-size: 16px;
    color: $dark-color;
    background-color: $light-color;
}

.container {
    margin: 2rem auto;
    max-width: 600px;
    padding: 1rem;
    background-color: $light-color;
    border-radius: 5px;
    box-shadow: 0 0 10px rgba(0, 0, 0, 0.2);
}

.text-center {
    text-align: center;
}

.list-group {
    margin-top: 1rem;
}

.todo-header {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-bottom: 1rem;

    &__title {
        font-size: 2rem;
        font-weight: bold;
        color: $primary-color;
    }

    &__delete-all-button {
        border: none;
        background-color: $danger-color;
        color: $light-color;
        padding: 0.5rem 1rem;
        border-radius: 5px;
        font-size: 1rem;
        cursor: pointer;
        transition: background-color 0.2s;

        &:hover {
            background-color: darken($danger-color, 10%);
        }
    }
}

.todo-item {
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 1rem;
    border-bottom: 1px solid $light-color;

    &:last-child {
        border-bottom: none;
    }

    &__task {
        font-size: 1.2rem;
        flex-grow: 1;
        text-decoration: none;
        color: $dark-color;
        transition: color 0.2s;

        &:hover {
            color: darken($dark-color, 10%);
        }

        &--completed {
            text-decoration: line-through;
            color: #a9a9a9;
        }
    }

    &__buttons {
        display: flex;
        align-items: center;

        button {
            border: none;
            border-radius: 5px;
            padding: 0.5rem 1rem;
            font-size: 1rem;
            cursor: pointer;
            transition: background-color 0.2s;

            &:hover {
                background-color: $light-color;
            }
        }

        &__edit-button {
            border: none;
            background-color: $primary-color;
            color: $light-color;
            margin-right: 0.5rem 1rem;
            border-radius: 5px;
            padding: 0.5rem 1rem;
            font-size: 1rem;
            cursor: pointer;

            &:hover {
                background-color: darken($primary-color, 10%);
            }
        }

        &__remove-button {
            border: none;
            background-color: $danger-color;
            color: $light-color;
            padding: 0.5rem 1rem;
            border-radius: 5px;
            font-size: 1rem;
            cursor: pointer;
            transition: background-color 0.2s;

            &:hover {
                background-color: darken($danger-color, 10%);
            }
        }

        &__toggle-completion-button {
            border: none;
            background-color: $success-color;
            color: $light-color;
            border-radius: 5px;
            padding: 0.5rem 1rem;
            font-size: 1rem;
            cursor: pointer;

            &:hover {
                background-color: darken($success-color, 10%);
            }
        }
    }
}

.new-todo-form {
    display: flex;
    margin-top: 1rem;

    input {
        flex-grow: 1;
        padding: 0.5rem;
        border: none;
        border-radius: 5px;
    }
}