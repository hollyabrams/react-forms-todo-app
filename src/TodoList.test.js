import { render } from '@testing-library/react';
import TodoList from './TodoList';

describe('TodoList component', () => {
  it('renders without crashing', () => {
    render(<TodoList />);
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
  });
});
