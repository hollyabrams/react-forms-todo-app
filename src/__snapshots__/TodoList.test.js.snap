// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="todo-header"
    >
      <h1
        class="todo-header__title"
      >
        Todo List
      </h1>
      <button
        class="todo-header__delete-all-button"
      >
        Delete All
      </button>
    </div>
    <form>
      <label
        for="task"
      >
        Task:
      </label>
      <input
        id="task"
        name="task"
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Add Todo
      </button>
    </form>
    <ul
      class="list-group"
    />
  </div>
</DocumentFragment>
`;
